services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/DownloadDirectoryHelperParameterizedTest.java

```diff
@@ -0,0 +1,177 @@
+import java.util.stream.Collectors;
+import org.assertj.core.util.Sets;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.ArgumentCaptor;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.testutils.FileUtils;
+import software.amazon.awssdk.transfer.s3.CompletedDirectoryDownload;
+import software.amazon.awssdk.transfer.s3.CompletedFileDownload;
+import software.amazon.awssdk.transfer.s3.DirectoryDownload;
+import software.amazon.awssdk.transfer.s3.DownloadDirectoryRequest;
+import software.amazon.awssdk.transfer.s3.DownloadFileRequest;
+import software.amazon.awssdk.transfer.s3.FileDownload;
+import software.amazon.awssdk.transfer.s3.internal.progress.DefaultTransferProgress;
+import software.amazon.awssdk.transfer.s3.internal.progress.DefaultTransferProgressSnapshot;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * Testing {@link DownloadDirectoryHelper} with different file systems.
+ */
+@RunWith(Parameterized.class)
+public class DownloadDirectoryHelperParameterizedTest {
+    private static final Set<Configuration> FILE_SYSTEMS = Sets.newHashSet(Arrays.asList(Configuration.unix(),
+                                                                                         Configuration.osX(),
+                                                                                         Configuration.windows()));
+    private Function<DownloadFileRequest, FileDownload> singleDownloadFunction;
+    private ListObjectsRecursivelyHelper listObjectsHelper;
+    private DownloadDirectoryHelper downloadDirectoryHelper;
+    private Path directory;
+    private static final String DIRECTORY_NAME = "test";
+
+    @Parameterized.Parameter
+    public Configuration configuration;
+
+    private FileSystem jimfs;
+
+    @Parameterized.Parameters
+    public static Collection<Configuration> fileSystems() {
+        return FILE_SYSTEMS;
+    }
+
+    @Before
+    public void methodSetup() throws IOException {
+        singleDownloadFunction = mock(Function.class);
+        listObjectsHelper = mock(ListObjectsRecursivelyHelper.class);
+        jimfs = Jimfs.newFileSystem(configuration);
+        downloadDirectoryHelper = new DownloadDirectoryHelper(TransferManagerConfiguration.builder().build(),
+                                                              jimfs,
+                                                              singleDownloadFunction,
+                                                              listObjectsHelper);
+        directory = jimfs.getPath("test");
+    }
+
+    @After
+    public void tearDown() {
+        if (jimfs != null) {
+            IoUtils.closeQuietly(jimfs, null);
+        } else {
+            FileUtils.cleanUpTestDirectory(directory);
+        }
+    }
+
+    @Test
+    public void downloadDirectory_shouldRecursivelyDownload() {
```

Comment:
Can we assert (either here or in integ tests) that the keys are visited in DFS order?